libevm/triedb/firewood/firewood.go

```diff
@@ -0,0 +1,518 @@
+
+	// proposalLock protects the proposal map and tree during updates.
+	proposalLock sync.RWMutex
+	// proposalMap provides O(1) access by state root to all proposals stored in the proposalTree
+	proposalMap map[common.Hash][]*proposal
+	// The proposal tree tracks the structure of the current proposals, and which proposals are children of which.
+	// This is used to ensure that we can dereference proposals correctly and commit the correct ones
+	// in the case of duplicate state roots.
+	// The root of the tree is stored here, and represents the top-most layer on disk.
+	proposalTree *proposal
+
+	// possibleProposals temporarily holds proposals created during a trie update.
+	// This is cleared after the update is complete and the proposals have been sent to the database.
+	possibleProposals []*proposal
+}
+
+type Config struct {
+	ChainDir             string
+	CleanCacheSize       int  // Size of the clean cache in bytes
+	FreeListCacheEntries uint // Number of free list entries to cache
+	Revisions            uint
+	ReadCacheStrategy    ffi.CacheStrategy
+}
+
+// Note that `FilePath` is not specified, and must always be set by the user.
+var Defaults = Config{
+	CleanCacheSize:       1024 * 1024, // 1MB
+	FreeListCacheEntries: 40_000,
+	Revisions:            100,
+	ReadCacheStrategy:    ffi.CacheAllReads,
+}
+
+func (c Config) BackendConstructor(ethdb.Database) triedb.DBOverride {
+	db, err := New(c)
+	if err != nil {
+		log.Crit("firewood: error creating database", "error", err)
+	}
+	return db
+}
+
+// New creates a new Firewood database with the given disk database and configuration.
+// Any error during creation will cause the program to exit.
+func New(config Config) (*Database, error) {
+	if err := validatePath(config.ChainDir); err != nil {
+		return nil, err
+	}
+
+	// Start the logs prior to opening the database
+	logPath := filepath.Join(config.ChainDir, logFileName)
+	if err := ffi.StartLogs(&ffi.LogConfig{Path: logPath}); err != nil {
+		// This shouldn't be a fatal error, as this can only be called once per thread.
+		// Specifically, this will return an error in unit tests.
+		log.Warn("firewood: error starting logs", "error", err)
+	}
+
+	dbPath := filepath.Join(config.ChainDir, dbFileName)
+	fw, err := ffi.New(dbPath, &ffi.Config{
+		NodeCacheEntries:     uint(config.CleanCacheSize) / 256, // TODO: estimate 256 bytes per node
+		FreeListCacheEntries: config.FreeListCacheEntries,
+		Revisions:            config.Revisions,
+		ReadCacheStrategy:    config.ReadCacheStrategy,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	currentRoot, err := fw.Root()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Database{
+		Firewood:    fw,
+		proposalMap: make(map[common.Hash][]*proposal),
+		proposalTree: &proposal{
+			info: &blockInfo{
+				root: common.Hash(currentRoot),
+				hashes: map[common.Hash]struct{}{
+					{}: {}, // genesis block has empty hash
+				},
+			},
+		},
+	}, nil
+}
+
+func validatePath(dir string) error {
```

Comment:
```suggestion
func validateDir(dir string) error {
```

(nit) Identifiers should communicate as much as possible about what they do.